File: java/subsets.java
class Solution {
    
    //store stack and solutions outside of recursion flow
    Stack<Integer> stack = new Stack();
    int[] nums = new int[0];
    List<List<Integer>> solutions = new ArrayList();
    
    public List<List<Integer>> subsets(int[] nums) {
        this.nums = nums;
        backtrack(0);
        return solutions;
    }
    
    private void backtrack(int index) {
        //if we've gone through the list of nums, add the present stack to the list of solutions
        if (index == nums.length) {
            solutions.add(new ArrayList(stack));
            return;
        }
        //call the next backtrack function both with nums[index] and without nums[index]
        backtrack(index + 1);
        stack.add(nums[index]);
        backtrack(index + 1);
        stack.pop();
    }
}

File: python/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        #store the stack and solutions outside of the recursion flow
        self.stack = []
        self.solutions = []
        self.nums = nums
        self.backtrack(0)
        return self.solutions
    
    def backtrack(self, index: int) -> None:
        #if we've gone through all of nums, add the present stack to the solutions list
        if index == len(self.nums):
            self.solutions.append(self.stack[:])
            return
        #otherwise, increment index and call the backtracking function with both the next number added or not added
        self.backtrack(index + 1)
        self.stack.append(self.nums[index])
        self.backtrack(index + 1)
        self.stack.pop()